File: frontend/src/lib/cnfDailyValues.ts
/**
 * Health Canada % Daily Value (%DV) reference, keyed by CNF NutrientID.
 *
 * The values now live in a single committed data file, cnfDailyValues.data.json,
 * which is byte-mirrored by backend/api/data/cnf_daily_values.json so the frontend
 * %DV display and the backend %DV-threshold filtering can never drift. A backend
 * parity test pins the two `values` maps equal.
 *
 * Source: Health Canada — "Nutrition labelling: Table of daily values" (canada.ca),
 * published 2022-10-20. In-force table (the 2016-12-14 version sunset 2025-12-31).
 * Adult / general-population column: Part 1 "Column 3" for macronutrients + sodium,
 * Part 2 "Column 4" ("any other case") for vitamins and minerals. Vitamin forms follow
 * Food and Drug Regulations D.01.003 (Vitamin A = RAE 320, Vitamin D = D2 + D3 µg 328,
 * Vitamin E = α-tocopherol 323, Niacin = NE 409, Folate = DFE 435).
 *
 * Deliberately NOT included: cholesterol (601, no %DV on Canadian labels), protein and
 * total carbohydrate (no Canadian %DV), iodide / chromium / molybdenum / chloride
 * (absent from CNF). See the data file's _provenance for the full note.
 *
 * %DV is computed against the per-100 g value (or per serving, if scaled upstream):
 *   %DV = nutrient_amount / daily_value × 100
 * The NutrientID fixes the unit, so the amount and the DV are always in the same unit.
 */
import dvData from './cnfDailyValues.data.json';

export interface CnfDailyValue {
  /** Daily Value amount, in `unit`. */
  dv: number;
  /** Unit the DV (and the matching CNF nutrient) is expressed in. */
  unit: 'g' | 'mg' | 'µg';
  /** Short label for tooltips. */
  label: string;
  /**
   * Saturated fat's %DV is computed against the sum of saturated + trans fat.
   * On NutrientID 606 (saturated) this points to 605 (trans) so the numerator is summed.
   */
  sumWithNutrientId?: number;
}

interface RawDailyValue {
  dv: number;
  unit: string;
  label: string;
  sum_with_nutrient_id?: number;
}

export const CNF_DAILY_VALUES: Record<number, CnfDailyValue> = Object.fromEntries(
  Object.entries(dvData.values as Record<string, RawDailyValue>).map(([id, v]) => [
    Number(id),
    {
      dv: v.dv,
      unit: v.unit as CnfDailyValue['unit'],
      label: v.label,
      ...(v.sum_with_nutrient_id != null ? { sumWithNutrientId: v.sum_with_nutrient_id } : {}),
    },
  ]),
);

/**
 * Compute %DV for one nutrient amount, or null when the nutrient has no Health Canada DV.
 * `lookupOtherById` supplies the trans-fat value when summing saturated + trans.
 */
export function percentDV(
  nutrientId: number,
  amount: number,
  lookupOtherById?: (nutrientId: number) => number | null,
): number | null {
  const entry = CNF_DAILY_VALUES[nutrientId];
  if (!entry || !(entry.dv > 0) || !Number.isFinite(amount)) return null;
  let numerator = amount;
  if (entry.sumWithNutrientId != null && lookupOtherById) {
    const other = lookupOtherById(entry.sumWithNutrientId);
    if (other != null && Number.isFinite(other)) numerator += other;
  }
  return (numerator / entry.dv) * 100;
}

File: backend/api/services/cnf_daily_values.py
"""Health Canada % Daily Value (%DV) reference, loaded from the committed JSON.

api/data/cnf_daily_values.json is the canonical table; it is byte-mirrored by the
frontend's cnfDailyValues.data.json so %DV display (frontend) and %DV-threshold
filtering (this module, used by the nutrient-discovery endpoint) can never drift. A
parity test pins the two `values` maps equal.

Keyed by CNF NutrientID; each id fixes the unit, so an amount and its DV are always in
the same unit. %DV = amount / dv * 100, computed on the per-100 g amount. Saturated fat
(606) sums with trans fat (605) via sum_with_nutrient_id.
"""
from __future__ import annotations

import json
import logging
import threading
from pathlib import Path
from typing import Callable, Dict, Optional

logger = logging.getLogger(__name__)

_PATH = Path(__file__).resolve().parent.parent / 'data' / 'cnf_daily_values.json'

_lock = threading.Lock()
_cache: Optional[Dict[int, Dict]] = None


def _load() -> Dict[int, Dict]:
    if not _PATH.exists():
        logger.warning('No daily-values file at %s; %%DV unavailable', _PATH)
        return {}
    try:
        raw = json.loads(_PATH.read_text(encoding='utf-8'))
    except Exception as exc:  # noqa: BLE001
        logger.warning('Daily-values file unreadable: %s', exc)
        return {}
    out: Dict[int, Dict] = {}
    for nid_str, rec in raw.get('values', {}).items():
        try:
            nid = int(nid_str)
        except (TypeError, ValueError):
            continue
        out[nid] = {
            'dv': float(rec['dv']),
            'unit': str(rec.get('unit', '')),
            'label': str(rec.get('label', '')),
            'sum_with_nutrient_id': (int(rec['sum_with_nutrient_id'])
                                     if rec.get('sum_with_nutrient_id') is not None else None),
        }
    return out


def _ensure() -> None:
    global _cache
    if _cache is not None:
        return
    with _lock:
        if _cache is None:
            _cache = _load()


def get_daily_value(nutrient_id: int) -> Optional[Dict]:
    """Return {dv, unit, label, sum_with_nutrient_id} for a nutrient, or None if it has no DV."""
    _ensure()
    assert _cache is not None
    return _cache.get(int(nutrient_id))


def has_daily_value(nutrient_id: int) -> bool:
    return get_daily_value(nutrient_id) is not None


def all_daily_values() -> Dict[int, Dict]:
    """Full {nutrient_id: {dv, unit, label, sum_with_nutrient_id}} table."""
    _ensure()
    assert _cache is not None
    return dict(_cache)


def percent_dv(
    nutrient_id: int,
    amount: float,
    lookup_other: Optional[Callable[[int], Optional[float]]] = None,
) -> Optional[float]:
    """%DV for one amount, or None when the nutrient has no DV. `lookup_other` supplies
    the trans-fat amount when summing saturated + trans (NutrientID 606)."""
    entry = get_daily_value(nutrient_id)
    if entry is None or not (entry['dv'] > 0):
        return None
    try:
        numerator = float(amount)
    except (TypeError, ValueError):
        return None
    other_id = entry.get('sum_with_nutrient_id')
    if other_id is not None and lookup_other is not None:
        other = lookup_other(int(other_id))
        if other is not None:
            try:
                numerator += float(other)
            except (TypeError, ValueError):
                pass
    return numerator / entry['dv'] * 100.0


def reset_for_test() -> None:
    global _cache
    with _lock:
        _cache = None
